src/passes/OptimizeInstructions.cpp

```diff
@@ -1345,15 +1377,57 @@ struct OptimizeInstructions
         }
       }
       // operations on all 1s
-      // TODO: shortcut method to create an all-ones?
       if (right->value == Literal(int32_t(-1)) ||
           right->value == Literal(int64_t(-1))) {
         if (binary->op == Abstract::getBinary(type, Abstract::And)) {
+          // x & -1   ==>   x
           return binary->left;
         } else if (binary->op == Abstract::getBinary(type, Abstract::Or) &&
                    !EffectAnalyzer(getPassOptions(), features, binary->left)
                       .hasSideEffects()) {
+          // x & -1   ==>   -1
           return binary->right;
+        } else if (binary->op == Abstract::getBinary(type, Abstract::Sub)) {
+          // x - (-1)   ==>   x + 1
+          Builder builder(*getModule());
+          return builder.makeBinary(
+            Abstract::getBinary(type, Abstract::Add),
+            binary->left,
+            builder.makeConst(Literal::makeFromInt32(1, type)));
+        } else if ((binary->op == Abstract::getBinary(type, Abstract::RemS) ||
+                    binary->op == GtUInt32 || binary->op == GtUInt64) &&
+                   !EffectAnalyzer(getPassOptions(), features, binary->left)
+                      .hasSideEffects()) {
+          // (signed)x % -1     ==>   0
+          // (unsigned)x > -1   ==>   0
+          return LiteralUtils::makeZero(type, *getModule());
+        } else if (binary->op == LtUInt32 || binary->op == LtUInt64 ||
+                   binary->op == Abstract::getBinary(type, Abstract::DivU)) {
+          // (unsigned)x / -1   ==>   x != -1
+          // (unsigned)x < -1   ==>   x != -1
+          return Builder(*getModule())
+            .makeBinary(Abstract::getBinary(type, Abstract::Ne),
+                        binary->left,
+                        binary->right);
+        } else if (binary->op == Abstract::getBinary(type, Abstract::Mul)) {
+          // (signed)x * -1   ==>   -x
+          Builder builder(*getModule());
+          return builder.makeBinary(
+            Abstract::getBinary(type, Abstract::Sub),
+            builder.makeConst(Literal::makeFromInt32(0, type)),
+            binary->left);
+        } else if ((binary->op == LeUInt32 || binary->op == LeUInt64) &&
+                   !EffectAnalyzer(getPassOptions(), features, binary->left)
+                      .hasSideEffects()) {
+          // (unsigned)x <= -1   ==>   1
+          return LiteralUtils::makeFromInt32(1, type, *getModule());
+        } else if (binary->op == LeSInt32 || binary->op == LeSInt64) {
+          // (signed)x <= -1   ==>   (unsigned)x >> sizeof(bits) - 1
```

Comment:
comment should mention why this is worth it, as the binary size is the same.

the constant would be smaller and perhaps more likely to compress, although -1 is pretty common too?

Reply:
I just remove this for now